services/galley/src/Galley/API/Update.hs

```diff
@@ -679,6 +684,53 @@ checkReusableCode convCode = do
   mapErrorS @'GuestLinksDisabled @'CodeNotFound $
     Query.ensureGuestLinksEnabled @db (Data.convTeam conv)
 
+updateConversationProtocolWithLocalUser ::
+  forall r.
+  ( Member (ErrorS 'ConvNotFound) r,
+    Member (ErrorS 'ConvInvalidProtocolTransition) r,
+    Member (ErrorS 'ConvMemberNotFound) r,
+    Member (Error FederationError) r,
+    Member MemberStore r,
+    Member ConversationStore r
+  ) =>
+  Local UserId ->
+  ClientId ->
+  ConnId ->
+  Qualified ConvId ->
+  ProtocolUpdate ->
+  Sem r ()
+updateConversationProtocolWithLocalUser lusr client conn qcnv update =
+  foldQualified
+    lusr
+    (\lcnv -> updateLocalConversationProtocol (tUntagged lusr) client (Just conn) lcnv update)
+    (\_rcnv -> throw FederationNotImplemented)
+    qcnv
+
+updateLocalConversationProtocol ::
+  forall r.
+  ( Member (ErrorS 'ConvNotFound) r,
+    Member (ErrorS 'ConvInvalidProtocolTransition) r,
+    Member (ErrorS 'ConvMemberNotFound) r,
+    Member MemberStore r,
+    Member ConversationStore r
+  ) =>
+  Qualified UserId ->
+  ClientId ->
+  Maybe ConnId ->
+  Local ConvId ->
+  ProtocolUpdate ->
+  Sem r ()
+updateLocalConversationProtocol qusr client _mconn lcnv (ProtocolUpdate newProtocol) = do
+  conv <- E.getConversation (tUnqualified lcnv) >>= noteS @'ConvNotFound
+  void $ ensureOtherMember lcnv qusr conv
+  case (protocolTag (convProtocol conv), newProtocol) of
+    (ProtocolProteusTag, ProtocolMixedTag) -> do
+      E.updateToMixedProtocol lcnv MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519
+      E.addMLSClients (convToGroupId lcnv) qusr (Set.singleton (client, nullKeyPackageRef))
+    (ProtocolMixedTag, ProtocolMixedTag) ->
+      pure ()
+    (_, _) -> throwS @'ConvInvalidProtocolTransition
```

Comment:
For consistency, maybe we can accept transitions from protocol `p` to `p`.

Reply:
Makes sense. I've added it